drivers/dai/nxp/sai.h

```diff
@@ -0,0 +1,518 @@
+	uint32_t bclk_rate;
+
+	/* TDM-related data */
+	uint32_t tdm_slots;
+	uint32_t rx_slots;
+	uint32_t tx_slots;
+	uint16_t tdm_slot_width;
+	uint16_t reserved2;
+};
+
+#ifdef CONFIG_SAI_HAS_MCLK_CONFIG_OPTION
+static int get_msel(sai_bclk_source_t bclk_source, uint32_t *msel)
+{
+	switch (bclk_source) {
+	case kSAI_BclkSourceMclkOption1:
+		*msel = 0;
+		break;
+	case kSAI_BclkSourceMclkOption2:
+		*msel = (0x2 << SAI_MCLK_MCR_MSEL_SHIFT);
+		break;
+	case kSAI_BclkSourceMclkOption3:
+		*msel = (0x3 << SAI_MCLK_MCR_MSEL_SHIFT);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void set_msel(uint32_t regmap, int msel)
+{
+	UINT_TO_I2S(regmap)->MCR &= ~SAI_MCLK_MCR_MSEL_MASK;
+	UINT_TO_I2S(regmap)->MCR |= msel;
+}
+
+static int clk_lookup_by_name(const struct sai_clock_data *clk_data, char *name)
+{
+	int i;
+
+	for (i = 0; i < clk_data->clock_num; i++) {
+		if (!strcmp(name, clk_data->clock_names[i])) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int get_mclk_rate(const struct sai_clock_data *clk_data,
+			 sai_bclk_source_t bclk_source,
+			 uint32_t *rate)
+{
+	int clk_idx;
+	char *clk_name;
+
+	switch (bclk_source) {
+	case kSAI_BclkSourceMclkOption1:
+		clk_name = "mclk1";
+		break;
+	case kSAI_BclkSourceMclkOption2:
+		clk_name = "mclk2";
+		break;
+	case kSAI_BclkSourceMclkOption3:
+		clk_name = "mclk3";
+		break;
+	default:
+		LOG_ERR("invalid bitclock source: %d", bclk_source);
+		return -EINVAL;
+	}
+
+	clk_idx = clk_lookup_by_name(clk_data, clk_name);
+	if (clk_idx < 0) {
+		LOG_ERR("failed to get clock index for %s", clk_name);
+		return clk_idx;
+	}
+
+	return clock_control_get_rate(clk_data->dev,
+				      UINT_TO_POINTER(clk_data->clocks[clk_idx]),
+				      rate);
+}
+#endif /* CONFIG_SAI_HAS_MCLK_CONFIG_OPTION */
+
+static inline void get_bclk_default_config(sai_bit_clock_t *cfg)
+{
+	memset(cfg, 0, sizeof(sai_bit_clock_t));
+
+	/* by default, BCLK has the following properties:
+	 *
+	 *	1) BCLK is active HIGH.
+	 *	2) BCLK uses MCLK1 source. (only applicable to master mode)
+	 *	3) No source swap.
+	 *	4) No input delay.
+	 */
+	cfg->bclkPolarity = kSAI_PolarityActiveHigh;
+	cfg->bclkSource = kSAI_BclkSourceMclkOption1;
+}
+
+static inline void get_fsync_default_config(sai_frame_sync_t *cfg)
+{
+	memset(cfg, 0, sizeof(sai_frame_sync_t));
+
+	/* by default, FSYNC has the following properties:
+	 *
+	 *	1) FSYNC is asserted one bit early with respect to the next
+	 *	frame.
+	 *	2) FSYNC is active HIGH.
+	 */
+	cfg->frameSyncEarly = true;
+	cfg->frameSyncPolarity = kSAI_PolarityActiveHigh;
+}
+
+static inline void get_serial_default_config(sai_serial_data_t *cfg)
+{
+	memset(cfg, 0, sizeof(sai_serial_data_t));
+
+	/* by default, the serial configuration has the following quirks:
+	 *
+	 * 1) Data pin is not tri-stated.
+	 * 2) MSB is first.
+	 */
+	/* note: this is ignored if xCR4 doesn't have CHMOD */
+#ifdef CONFIG_SAI_HAS_XCR4_CHMOD
+	cfg->dataMode = kSAI_DataPinStateOutputZero;
```

Comment:
> That's probably because only 8MP (MIMX8ML8) has CHMOD field in XCR4, right?

As well as i.MX93, yes.



> Maybe CONFIG_SAI_HAS_XCR4_CHMOD should be linked with FSL_FEATURE_SAI_HAS_CHANNEL_MODE?

This should work, yes. Will give it a try and update this PR if it works ok.